### PR TITLE
Parallel executor

### PR DESCRIPTION
**Pull request by avijit-nervana**

### CODE REVIEW

File: ngraph_bridge/ngraph_encapsulate_op.cc [hunk 1]
```diff
@@ -60,11 +60,114 @@ namespace tensorflow {
...
+  // Set the backend type for the this NGraphEncapsulate Op
+  std::string backend_name;
+  OP_REQUIRES_OK(ctx, ctx->GetAttr<string>("ngraph_backend", &backend_name));
+  std::string device_id;
+  OP_REQUIRES_OK(ctx, ctx->GetAttr<string>("ngraph_device_id", &device_id));
+
+  // Concatenate the backend_name:device_id
+  string be_name;
+  try {
+    be_name = BackendManager::GetBackendCreationString(backend_name, device_id);
```
Comment (shresthamalik) [inline ngraph_bridge/ngraph_encapsulate_op.cc]: BackendManager::GetBackendCreationString(backend_name, device_id) will not throw an error
so we dont need a try catch here
https://github.com/tensorflow/ngraph-bridge/blob/master/ngraph_bridge/ngraph_backend_manager.cc#L195

File: ngraph_bridge/ngraph_executor.cc [hunk 2]
```diff
@@ -0,0 +1,550 @@
...
+    if (std::getenv("NGRAPH_ENABLE_SERIALIZE") != nullptr) {
+      std::string file_name = "tf_function_" + m_name + ".json";
+      NgraphSerialize("tf_function_" + m_name + ".json", ng_function);
+#if defined NGRAPH_DISTRIBUTED
+      int rank_id;
+      rank_id = ng::get_distributed_interface()->get_rank();
+      NgraphSerialize(
+          "tf_function_" + m_name + "_" + to_string(rank_id) + ".json",
+          ng_function);
+#endif
```
Comment (shresthamalik) [inline ngraph_bridge/ngraph_executor.cc]: free this function of input output caches.  Is this in scope of this PR?

File: ngraph_bridge/ngraph_executor.h [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+
+  // ng_function, ng_executable, Output and Input Cache maps
+  std::unordered_map<std::string, std::shared_ptr<ngraph::runtime::Executable>>
+      m_ng_exec_map;
+  std::unordered_map<std::shared_ptr<ngraph::runtime::Executable>,
+                     std::shared_ptr<ngraph::Function>>
+      m_ng_function_map;
+
+  NgFunctionIOCache m_ng_exec_input_cache_map;
+  NgFunctionIOCache m_ng_exec_output_cache_map;
```
Comment (shresthamalik) [inline ngraph_bridge/ngraph_executor.h]: Remove the maps

File: ngraph_bridge/ngraph_executor.h [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+
+  // Get tensorflow input tensors, input shapes, static_inputs to Compute
+  // Signature
+  Status ComputeSignature(const std::vector<Tensor>& tf_input_tensors,
+                          std::vector<TensorShape>& input_shapes,
+                          std::vector<const Tensor*>& static_input_map,
+                          std::stringstream& signature_ss) const;
+
+ private:
+  const int my_instance_id;
```
Comment (shresthamalik) [inline ngraph_bridge/ngraph_executor.h]: should we call it m_instance_id;

File: ngraph_bridge/ngraph_executor.cc [hunk 5]
```diff
@@ -0,0 +1,550 @@
...
+  return Status::OK();
+}
+
+//---------------------------------------------------------------------------
+//  NGraphExecutor::GetNgExecutable
+//---------------------------------------------------------------------------
+Status NGraphExecutor::GetNgExecutable(
+    const std::vector<Tensor>& tf_input_tensors,
+    std::shared_ptr<ngraph::runtime::Executable>& ng_exec, bool& cache_hit) {
+  // FIRST Compute the signature
```
Comment (kanvi-nervana) [inline ngraph_bridge/ngraph_executor.cc]: Is this comment meant to be here

File: test/graph_rewrites/config_for_grappler_test.cc [hunk 6]
```diff
@@ -274,13 +273,17 @@ TEST(GrapplerConfig, RConfig3) {
   ASSERT_OK(GetNodeAttr(ng_encap->attrs(), "ngraph_backend", &ng_backend));
   ASSERT_OK(GetNodeAttr(ng_encap->attrs(), "ngraph_device_id", &ng_device_id));
 
-  ASSERT_EQ(ng_backend, "CPU");
+  // Even though the backend is set via config-writer, the one specified
+  // by the env. var takes effect. So though we set this to CPU
+  // the packend should point to NOP as set vis env. var
```
Comment (kanvi-nervana) [inline test/graph_rewrites/config_for_grappler_test.cc]:
```suggestion
  // the backend should point to NOP as set via env. var
```

File: ngraph_bridge/ngraph_executor.h (hunk 3 above)
Comment (sindhu-nervana) [reply ngraph_bridge/ngraph_executor.h]: +1

File: test/test_parallel_executor.cpp [hunk 7]
```diff
@@ -0,0 +1,576 @@
+/*******************************************************************************
+ * Copyright 2017-2019 Intel Corporation
```
Comment (kanvi-nervana) [inline test/test_parallel_executor.cpp]:
```suggestion
 * Copyright 2019 Intel Corporation
```

File: ngraph_bridge/ngraph_executor.cc (hunk 2 above)
Comment (avijit-nervana) [reply ngraph_bridge/ngraph_executor.cc]: Yes - but a different PR

File: ngraph_bridge/ngraph_executor.h [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+
+#include "logging/ngraph_log.h"
+#include "ngraph_bridge/ngraph_freshness_tracker.h"
+#include "ngraph_bridge/ngraph_pipelined_tensors.h"
+
+namespace tensorflow {
+
+namespace ngraph_bridge {
+
+using NgFunctionIOCache = std::unordered_map<
```
[the remark this answers is not in the document]
Comment (avijit-nervana) [reply ngraph_bridge/ngraph_executor.h]: Done.

File: ngraph_bridge/ngraph_encapsulate_op.cc [hunk 9]
```diff
@@ -60,11 +60,114 @@ namespace tensorflow {
...
+
+//---------------------------------------------------------------------------
+//  CreateParallelExecutor
+//---------------------------------------------------------------------------
+void NGraphEncapsulateOp::CreateParallelExecutor(OpKernelConstruction* ctx,
+                                                 const string& backend_name) {
+  GraphDef* graph_def;
+  unique_ptr<Graph> encap_subgraph(new Graph(OpRegistry::Global()));
+
+  int cluster{-1};
```
[the remark this answers is not in the document]
Comment (avijit-nervana) [reply ngraph_bridge/ngraph_encapsulate_op.cc]: Yes.

File: ngraph_bridge/ngraph_executor.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (avijit-nervana) [reply ngraph_bridge/ngraph_executor.h]: Will remove in a future PR.

File: ngraph_bridge/ngraph_encapsulate_op.cc [hunk 10]
```diff
@@ -60,11 +60,114 @@ namespace tensorflow {
...
+  OP_REQUIRES_OK(ctx, ctx->GetAttr<int>("ngraph_cluster", &cluster));
```
[the remark this answers is not in the document]
Comment (avijit-nervana) [reply ngraph_bridge/ngraph_encapsulate_op.cc]: Yes - a future PR.

File: examples/cpp/CMakeLists.txt [hunk 11]
```diff
@@ -14,5 +14,5 @@
 # limitations under the License.
 # ******************************************************************************
 message(STATUS "C++ Examples will be built")
-add_subdirectory(infer_multiple_networks)
-add_subdirectory(infer_single)
+#add_subdirectory(infer_multiple_networks)
+#add_subdirectory(infer_single)
```
Comment (shresthamalik) [inline examples/cpp/CMakeLists.txt]: Commented code. Should this be removed?

File: ngraph_bridge/ngraph_encapsulate_op.cc [hunk 12]
```diff
@@ -276,6 +381,154 @@ NGraphEncapsulateOp::~NGraphEncapsulateOp() {
...
+  event_compile.Stop();
+  ngraph::Event::write_trace(event_compile);
+
+  // Get the pipelned tensors
+  ngraph::Event event_get_tensor("Get Tensor", "", "");
+
+  std::tuple<int, PipelinedTensorVector, PipelinedTensorVector> io_tensors;
+  OP_REQUIRES_OK(
+      ctx, m_parallel_executor->GetTensorsFromPipeline(ng_exec, io_tensors));
+
```
Comment (shresthamalik) [inline ngraph_bridge/ngraph_encapsulate_op.cc]: do we need to check if the depth in io_tensors is not -1 before proceeding?
ie
get<0>(io_tensors)>=0

File: test/test_utilities.cpp [hunk 13]
```diff
@@ -240,6 +238,69 @@ bool Compare(float desired, float actual, float rtol, float atol) {
...
+  }
+  return Status::OK();
+}
+
+template <>
+void AssignInputValues(Tensor& A, int8 x) {
+  auto A_flat = A.flat<int8>();
+  auto A_flat_data = A_flat.data();
+  for (int i = 0; i < A_flat.size(); i++) {
+    cout << (int)x << " ";
```
Comment (shresthamalik) [inline test/test_utilities.cpp]: remove cout here

File: test/test_utilities.cpp [hunk 14]
```diff
@@ -240,6 +238,69 @@ bool Compare(float desired, float actual, float rtol, float atol) {
...
+  tensorflow::GraphDef graph_def;
+  auto load_graph_status =
+      ReadTextProto(Env::Default(), graph_file_name, &graph_def);
+  if (!load_graph_status.ok()) {
+    return tensorflow::errors::NotFound("Failed to load compute graph at '",
+                                        graph_file_name, "'");
+  }
+  session->reset(tensorflow::NewSession(options));
+  Status session_create_status = (*session)->Create(graph_def);
+  if (!session_create_status.ok()) {
```
Comment (shresthamalik) [inline test/test_utilities.cpp]: we can return session_create_status

File: ngraph_bridge/thread_safe_queue.h [hunk 15]
```diff
@@ -0,0 +1,61 @@
...
+      m_cv.Wait(&m_mutex);
+    }
+
+    next = m_queue.front();
+    m_queue.pop();
+    m_mutex.Unlock();
+    return next;
+  }
+
+  void Add(T* session) {
```
Comment (shresthamalik) [inline ngraph_bridge/thread_safe_queue.h]: It's a generic queue
Rename session to object

File: examples/cpp/CMakeLists.txt (hunk 11 above)
Comment (avijit-nervana) [reply examples/cpp/CMakeLists.txt]: Done. Reorganized the cpp directory to have a new `inference` directory with the contents of test/benchmark moved.

File: ngraph_bridge/ngraph_encapsulate_op.cc (hunk 12 above)
Comment (avijit-nervana) [reply ngraph_bridge/ngraph_encapsulate_op.cc]: The check it's already done by the GetTensorsFromPipeline()

File: ngraph_bridge/ngraph_executor.cc [hunk 16]
```diff
@@ -0,0 +1,552 @@
...
+  NGRAPH_VLOG(3) << "NGraphExecutor(): " << instance_id
+                 << " Backend: " << backend_name;
+
+  // Get the backend. Note that the backend may not be available
+  // so that's a programmng error.
+  try {
+    auto backend = BackendManager::GetBackend(m_op_backend_name);
+    m_executable_can_create_tensor = backend->executable_can_create_tensors();
+  } catch (...) {
+    throw std::runtime_error("No backend available. Cannot execute graph");
```
[the remark this answers is not in the document]
Comment (avijit-nervana) [reply ngraph_bridge/ngraph_executor.cc]: @sayantan-nervana this is wrong suggestion - the code won't compile!

[the remark this answers is not in the document]
Comment (sayantan-nervana) [reply ngraph_bridge/ngraph_executor.cc]: ah I read it as TF error status, which concats strings with commas. This is `runtime_error`, so we need `+`. My point is we should print the backend name `m_op_backend_name`

File: examples/cpp/inference/infer_single_network.cc [hunk 17]
```diff
@@ -0,0 +1,274 @@
...
+//-----------------------------------------------------------------------------
+//  The benchmark test for inference does the following
+//    1. Preloads the input image buffer (currently single image)
+//    2. Creates the TensorFlow Session by loading a frozem inference graph
+//    3. Starts the worker threads and runs the test for a specifed iterations
+//
+//  Each worker thread does the following:
+//    1. Gets an image from the image pool
+//    2. Copies the data to a TensorFlow Tensor
+//    3. Runs the inference using the same session used by others as well
```
[the remark this answers is not in the document]
Comment (avijit-nervana) [reply examples/cpp/inference/infer_single_network.cc]: This directory will be removed.

File: ngraph_bridge/ngraph_executor.cc [hunk 18]
```diff
@@ -0,0 +1,553 @@
...
+                                                 m_graph.get(), ng_function));
+      ng_function->set_friendly_name(m_node_name);
+    } else {
+      auto itr = m_aot_functions.find(signature);
+      if (itr == m_aot_functions.end()) {
+        return errors::Internal(
+            "Expected to find AOT precompiled ng function of signature: ",
+            signature);
+      }
+      ng_function = ng::deserialize(itr->second);
```
Comment (sayantan-nervana) [inline ngraph_bridge/ngraph_executor.cc]: We are deserializing the string here. We do not need to deserialize (and deserializing will break networks like ssd).

In impl, we do not cache the ng_function at all: https://github.com/tensorflow/ngraph-bridge/blob/master/ngraph_bridge/ngraph_encapsulate_impl.cc#L151

we just have a map of exec->string (where string is the serialized function) rather than exec->function

Comment (avijit-nervana) [reply ngraph_bridge/ngraph_executor.cc]: Please fix this once my PR is in.

File: examples/cpp/inference/infer_single_network.cc [hunk 19]
```diff
@@ -0,0 +1,274 @@
...
+  // Instantiate the Engine
+  benchmark::InferenceEngine inference_engine("Foo");
+  TF_CHECK_OK(inference_engine.LoadImage(
+      graph, image_files, input_width, input_height, input_mean, input_std,
+      input_layer, output_layer, use_NCHW, preload_images, input_channels));
+
+  string backend_name = "CPU";
+  if (std::getenv("NGRAPH_TF_BACKEND") != nullptr) {
+    backend_name = std::getenv("NGRAPH_TF_BACKEND");
+  }
```
[the remark this answers is not in the document]
Comment (avijit-nervana) [reply examples/cpp/inference/infer_single_network.cc]: In a future PR.
